### ISSUE TITLE
Bug: [no-unused-vars] flags variables that can't be declared as a type-only

### ISSUE THREAD

**Issue by JacobLey** [typescript-eslint/typescript-eslint#10184]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.ts&code=MYewdgzgLgBApgDwA52FAKgTxQeQGYwC8MAPAIIB8AFAJREUxUDeAUDDFCAKICOArgEMANgC5GAsWTqEGAgDQsAvjQDcLFqEiwARgIBORRtIYBGAEwBmNZugw8IEId17aalohRosuPCSjY4EAJ7EGoaADpOXkEhKnMLVSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const expectTypeOf = <A>() => ({
  toEqual: (a: A) => a,
});

const bar = () => 123;
const foo = bar();

expectTypeOf<typeof foo>().toEqual(123);
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

### tsconfig

```jsonc
{}
```

### Expected Result

I expected that the `foo` variable is not flagged, since there is no way to capture the result of _actually_ running `bar()` without assigning to a variable, despite acknowledging the fact that the variable is not "used" any more.

### Actual Result

> 'foo' is assigned a value but only used as a type. 6:7 - 6:10

The variable is not used _in javascript_ beyond assignment, so the rule flags the variable.

### Additional Info

I will preface this bug report with the acknowledgement that the lint rule is probably _technically_ correct. In the example above, the `foo` variable is only every assigned and then not used at all _in javascript_. `foo` only serves a purpose as a type.

However I intentionally have done this because:
* I actually want to run the `bar()` method
  * Even if just for type convenience, it may be necessary to format code like that in order to properly apply generics
* I want to assert some feature in the result that can be done purely with types

I alluded to this a bit in the example (with an extremely over-simplified implementation), but a common use case for me is passing types during tests to the [expect-type](https://www.npmjs.com/package/expect-type) package. Sometimes the type is not the straightforward `foo` value, but a type modification like `Bar<foo>`, so I can't just supply it value.

An example in my actual codebase: https://github.com/JacobLey/leyman/blob/fedc4aefac23a2a3f0fb088a878da251fab32324/apps/juniper/src/tests/unit/lib/schema.spec.ts#L157
> 'json' is assigned a value but only used as a type

Where I am applying a `SchemaType<typeof json>` before comparing.

So admittedly the generated javascript looks like:
```ts
const bar = () => 123;
const foo = bar();

expectTypeOf().toEqual(123);
```
and _technically_ the "unused variable" rule is correct.

But _to my knowledge_ there is no way to achieve the logic I was trying to achieve without assigning to a variable.

i.e. The following is impossible/illegal in typescript
```ts
type foo = bar();

expectTypeOf<foo>().toEqual(123);
```

If somehow that _is_ possible I would love to hear about it, but otherwise would request that variables that still used in typescript types are allowed under this rule.

**Comment by kirkwaiblinger** [comment]
HI @JacobLey!

Noting for discussion that this is very similar to [no-unnecessary-type-parameters' inapplicability for type testing](https://typescript-eslint.io/rules/no-unnecessary-type-parameters/#when-not-to-use-it) and the discussion about no-unused-expressions in #9275.

**Comment by kirkwaiblinger** [comment]
I'm tentatively in favor of creating an option for both this case and for #9275

**Comment by JacobLey** [comment]
Another example I ran into is a place where I `declare`d a class for type-only purposes.
TLDR: I needed to infer the instance of a class even when the class has a private constructor. Its possible with some... hacks: [playground link](https://typescript-eslint.io/play#ts=5.6.3&fileType=.tsx&code=CYUwxgNghgTiAEUBGBnALjKY30lFK8AwgPYB26MArtshAgN4BQ88ADjAJYBuUaCYcpRpoSMABQA6abADmKAFzwqZANZkSAdzIBtALoBKANxMAvkwD0AKissr8AIIQAtiXTw0ACwSv3skGQgXGC40ATsbiicSPSSdnbwAAIoIIyJEJxq8J5oaGyKFhaynF5USJKCzhbOnGAwbiQAZmgWACoAnmwgAMp1nGwtnARUICgWAKwA7AAsAIwATKbwmp7tHp5D8JuBYKMosO3x8FYWTKB4cIioGFg4eOEACvX82CDApBQYInSMLOzP4H4wFwQi%2B2DEUhkMHkShU6i0ukMJnMaE6CAAkigiGFCABeDxoprEUHUWgxBAAHwJXSJTxILyBH2EZPoJlRXXg6M%2BUDIuwA8o1sfgUAAeVrwEAAD34ZGAhExQoIAD54PiuegebsOl0RX9CvA%2BTBQDB4ABZPj8GAoACEf3EgU08HEBlVKsC3CCLoAZPBWkwlSYmFK2GI7jj4AAxEgkeDMVhsMoZEJITKyVXwBYAZhM8a4vH4IM%2BpNEJudsfg5nMwdDoWF8AAQrBY38EzFavAkFAAF7pgDkyDAvZz-3pgLeheZJbLDArZiDkpDMBw7IQUZI6rQmoQau5vJAAsVopXRLXAfni%2BXaIbsA3W-Tt73B5xIuPjWvMADQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge3oA99oBDJog6xktMvH7JUGSImjQO0SABpw2PEVLIK1Og2ZtOtWE1GIAJrQBu-aDKjzFy9QF8QroA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

I am also getting flagged for no-unused-vars here, despite that `declare` means the variable doesn't exist at runtime (at least not explicitly).

It is also getting flagged for extraneous class (decided to keep that error in just for the record) but I don't personally consider that a bug (I'm happy `eslint-disable`-ing that line)

Example in actual codebase: https://github.com/JacobLey/leyman/blob/194f88d69e4c1bc7d1dd5696736357765caa1b18/tools/haywire/src/lib/types.ts#L6-L26

Perhaps this should be its own issue (my original callout was _actual_ variables where the rule is technically correct but unfeasible to workaround) but I suspect both were caused by the same feature where this rule is now being (IMHO) overly strict.

**Comment by bradzacher** [comment]
It is the rare edge case that you actually 100% need to have a value that's defined at runtime but only used at type time.

For most cases [you can use the "expression with type arguments" syntax](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAtgUwE4HMEGUpIJZhTAXhgB4BBGBADygTABMIZoc8AaGAIQutoaa1woAfAAoAhgC4YpdgCMpHAJRSABgBIA3qQC+EzR20rCQmOo1jdm2YYDcAKDuhIsAGYgQheMjSYWKEQDkYrLAAewBCC4oAYr2bh4QYlDYEC7YCIxBIRKR0fZ2UACeAA4InGJIngBKCFAArkhgAColCMRFpSAuXqgYAnjEWaHhdJEBQkL2APRTMHMwAHoA-A5AA) to supply the generics in conjunction with `ReturnType` to "call" the function in a type-only way.

From my anecdotal experience the remaining cases that you do need it are either
- "type only tests" where you're truly wanting to do some checks of the type using a type testing library, 
- edge-case hacks around TS itself.

As an example - I recently updated my company's codebase (~90k files) and there were only a few dozen violations. Of those violation, a small fraction were valid - the rest were all bad code. Granted - we don't do any "type testing" in our codebase - but it is a good, real-world example.

**Comment by JacobLey** [comment]
> `typeof mergeString<'abc', 'def'>`

!!

Is this a new feature!? I didn't know this was possible!

Ok then I'll agree that actually handles my whole "calling function with generics to get the return type".

I think the issue of `declare`d classes getting flagged is potentially still a "bug"?

While I'm testing stuff out, I think I came upon another case where this rule is actually under-reporting.
When an "unused var" is used as the key in a type, it doesn't get reported.
[playground link](https://typescript-eslint.io/play/#ts=5.6.3&fileType=.tsx&code=PTAEDkHsBdQJwKYAdJ2ggJqAFAIwK6wDOAFpPgDZa4ICUAUAMaQB2RsAZpJKALygBlAJ4BbXJArYA5AEMptANz0QoAErJU6LDKKgEcOKhzMDCRtApCGzNrFwy4fQaPGSpueUoQAPFGlAAlizocBwyjAigAEJBAOagAN70oKAA2lyQALoA-ABcoACMAEwAzEop3kIAXvnQQkgIkByg9nBKAL70ymAA6qgA1rrQJJFEMiKRAO4yQqCTIyyg%2BERxoBhmFA4y0AGsRN2gAIIUw%2BSxJKAOsfgyuJagLJCTLTAXpORULZEyFBRPmPR1oxNohQDZ2KAMkV8vgWAEAI74UYuCRKIEgyLguwOaFLOGI5FiVH0Hx%2BWBBEJhCLROJFRLJNJQnL5YplBmVGqgOoNJotHEdIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

Although I think this also highlights when it over-reports for `declare`d variables. 

I'll often use this pattern in my codebase to create "tagged/branded types" using symbols that are declared (convenient way to guarantee uniqueness in typescript).

**Comment by JacobLey** [comment]
Either way I think the original bug report "flags variables that can't be declared as type-only" has been addressed with the ability to apply generics to the function.
So will close this issue, although maybe a footnote on the documentation page with examples of how to avoid these pitfalls might be helpful.

I'll defer to you all/maintainers if you want to create a separate issue for handling `declare`d variables as well as reporting values used as the key of types.
